diffrax/_solver/langevin_srk.py

```diff
@@ -0,0 +1,320 @@
+    BoolScalarLike,
+    DenseInfo,
+    RealScalarLike,
+)
+from .._local_interpolation import LocalLinearInterpolation
+from .._solution import RESULTS
+from .._term import (
+    AbstractTerm,
+    LangevinDiffusionTerm,
+    LangevinDriftTerm,
+    LangevinTuple,
+    LangevinX,
+    MultiTerm,
+    WrapTerm,
+)
+from .base import AbstractItoSolver, AbstractStratonovichSolver
+
+
+_ErrorEstimate = TypeVar("_ErrorEstimate", None, LangevinTuple)
+_LangevinArgs = tuple[LangevinX, LangevinX, Callable[[LangevinX], LangevinX]]
+
+
+def get_args_from_terms(
+    terms: MultiTerm[tuple[AbstractTerm[Any, RealScalarLike], AbstractTerm]],
+) -> _LangevinArgs:
+    drift, diffusion = terms.terms
+    if isinstance(drift, WrapTerm):
+        assert isinstance(diffusion, WrapTerm)
+        drift = drift.term
+        diffusion = diffusion.term
+
+    assert isinstance(drift, LangevinDriftTerm)
+    assert isinstance(diffusion, LangevinDiffusionTerm)
+    gamma = drift.gamma
+    u = drift.u
+    f = drift.grad_f
+    return gamma, u, f
+
+
+class AbstractCoeffs(eqx.Module):
+    dtype: AbstractVar[jnp.dtype]
+
+
+_Coeffs = TypeVar("_Coeffs", bound=AbstractCoeffs)
+
+
+class SolverState(eqx.Module, Generic[_Coeffs]):
+    h: RealScalarLike
+    taylor_coeffs: PyTree[_Coeffs, "LangevinX"]
+    coeffs: _Coeffs
+    rho: LangevinX
+    prev_f: LangevinX
+
+
+# CONCERNING COEFFICIENTS:
+# The coefficients used in a step of this SRK depend on
+# the time increment h, and the parameter gamma.
+# Assuming the modelled SDE stays the same (i.e. gamma is fixed),
+# then these coefficients must be recomputed each time h changes.
+# Furthermore, for very small h, directly computing the coefficients
+# via the function below can cause large floating point errors.
+# Hence, we pre-compute the Taylor expansion of the SRK coefficients
+# around h=0. Then we can compute the SRK coefficients either via
+# the Taylor expansion, or via direct computation.
+# In short the Taylor coefficients give a Taylor expansion with which
+# one can compute the SRK coefficients more precisely for a small h.
+
+
+class AbstractLangevinSRK(
+    AbstractStratonovichSolver[SolverState],
+    AbstractItoSolver[SolverState],
+    Generic[_Coeffs, _ErrorEstimate],
+):
+    """Abstract class for Stochastic Runge Kutta methods specifically designed
+    for Underdamped Langevin Diffusion of the form
+
+    $$d x_t = v_t dt$$
+
+    $$d v_t = - gamma v_t dt - u ∇f(x_t) dt + (2gammau)^(1/2) dW_t$$
+
+    where $v$ is the velocity, $f$ is the potential, $gamma$ and $u$ are the
+    friction and momentum parameters, and $W$ is a Brownian motion.
+
+    Solvers which inherit from this class include [`diffrax.ALIGN`][],
+    [`diffrax.ShOULD`][], and [`diffrax.QUIC_SORT`][].
+    """
+
+    term_structure = MultiTerm[tuple[LangevinDriftTerm, LangevinDiffusionTerm]]
+    interpolation_cls = LocalLinearInterpolation
+    taylor_threshold: RealScalarLike = eqx.field(static=True)
```

Comment:
(Also it should probably be an `AbstractVar` rather than a field? Unusual to need to mandate that something be a field on an abstract class, normally we only care that we can do lookup by name.)

Reply:
That makes sense. I'm still learning the conventions 😅